sdc/datatypes/hpat_pandas_series_functions.py

```diff
@@ -445,65 +448,131 @@ def hpat_pandas_series_value_counts(self, normalize=False, sort=True, ascending=
+
+            index_string_lengths = numpy.asarray([len(s) for s in sorted_unique_values])
+            index_total_chars = numpy.sum(index_string_lengths)
+            result_index = pre_alloc_string_array(len(sorted_unique_values), index_total_chars)
+            cp_str_list_to_array(result_index, sorted_unique_values)
+
+            if not dropna:
+                # set null bit for StringArray element corresponding to NaN element of initial Series
+                index_previous_nan_pos = unique_values_len - 1
+                for i in numpy.arange(unique_values_len):
+                    if indexes_order[i] == index_previous_nan_pos:
+                        str_arr_set_na(result_index, i)
+                        break
+
             sorted_value_counts = numpy.take(value_counts, indexes_order)
 
-            return pandas.Series(sorted_value_counts, index=sorted_unique_values, name=self._name)
+            return pandas.Series(sorted_value_counts, index=result_index, name=self._name)
 
         return hpat_pandas_series_value_counts_str_impl
 
-    def hpat_pandas_series_value_counts_number_impl(self, normalize=False, sort=True, ascending=False, bins=None, dropna=True):
-        unique_values = self.unique()
+    elif isinstance(self.dtype, types.Number):
 
-        if dropna:
-            nan_mask = numpy.isnan(unique_values)
-            unique_values = unique_values[~nan_mask]
-        #else:
-        # TODO: unique() can not handle numpy.nan because numpy.nan == numpy.nan is False
+        series_values_type = numba.typeof(self.dtype)
+        def hpat_pandas_series_value_counts_number_impl(self, normalize=False, sort=True, ascending=False, bins=None, dropna=True):
 
-        # TODO: not optimal
-        value_counts = numpy.array([numpy.sum(self._data == value) for value in unique_values])
+            value_counts_dict = Dict.empty(
+                key_type=series_values_type,
+                value_type=types.intp
+            )
 
-        # Series.unique() returns values in ascending order
-        indexes_order = numpy.arange(len(unique_values))
-        if sort:
-            indexes_order = value_counts.argsort()
-            if not ascending:
-                indexes_order = indexes_order[::-1]
+            zero_counts = 0
+            is_zero_found = False
+            for value in self._data:
+                if (dropna and numpy.isnan(value)):
+                    continue
+
+                # Pandas hash-based value_count_float64 function doesn't distinguish between
+                # positive and negative zeros, hence we count zero values separately and store
+                # as a key the first zero value found in the Series
+                if not value:
+                    zero_counts += 1
+                    if not is_zero_found:
+                        zero_value = value
+                        is_zero_found = True
+                    continue
+
+                if value in value_counts_dict:
```

Comment:
consider:
```python
value_counts_dict[value] = value_counts_dict.get(value, 0) + 1
```

Reply:
@AlexanderKalistratov Thank you! Great that it's supported in nopython mode.